receiver/dcgmreceiver/client.go

```diff
@@ -352,103 +344,108 @@ func setWatchesOnFields(logger *zap.Logger, deviceGroup dcgm.GroupHandle, fieldI
-		client.logger.Debugf("Polling DCGM daemon for GPU %d", gpuIndex)
-		retry := true
-		for i := 0; retry && i < client.maxRetries; i++ {
-			fieldValues, pollErr := dcgmGetLatestValuesForFields(gpuIndex, client.enabledFieldIDs)
-			client.logger.Debugf("Got %d field values", len(fieldValues))
-			if pollErr == nil {
-				gpuMetrics[gpuIndex], retry = client.appendMetrics(gpuMetrics[gpuIndex], gpuIndex, fieldValues)
-				if retry {
-					client.logger.Warnf("Retrying poll of DCGM daemon for GPU %d; attempt %d", gpuIndex, i+1)
-					time.Sleep(client.pollingInterval)
-					continue
-				}
-				client.logger.Debugf("Successful poll of DCGM daemon for GPU %d", gpuIndex)
-			} else {
-				msg := fmt.Sprintf("Unable to poll DCGM daemon for GPU %d on %s", gpuIndex, pollErr)
-				client.issueWarningForFailedQueryUptoThreshold(gpuIndex, "all-profiling-metrics", msg)
-				err.AddPartial(1, fmt.Errorf("%s", msg))
-			}
-		}
+// collect will poll dcgm for any new metrics, updating client.devices as appropriate
+// It returns the estimated polling interval.
+func (client *dcgmClient) collect() (time.Duration, error) {
+	client.logger.Debugf("Polling DCGM daemon for field values")
+	if len(client.enabledFieldIDs) == 0 {
+		// Make sure we don't try to scrape without a device group (since we don't construct one when there are no enabled fields).
+		return 0, nil
 	}
-
-	return gpuMetrics, err.Combine()
-}
-
-func (client *dcgmClient) appendMetrics(gpuMetrics []dcgmMetric, gpuIndex uint, fieldValues []dcgm.FieldValue_v1) (result []dcgmMetric, retry bool) {
-	retry = false
+	fieldValues, pollTime, err := dcgmGetValuesSince(client.deviceGroup, client.enabledFieldGroup, client.lastSuccessfulPoll)
+	if err != nil {
+		msg := fmt.Sprintf("Unable to poll DCGM daemon for on %s", err)
+		client.issueWarningForFailedQueryUptoThreshold("all-profiling-metrics", msg)
+		return 0, err
+	}
+	client.logger.Debugf("Got %d field values over %s", len(fieldValues), pollTime.Sub(client.lastSuccessfulPoll))
+	client.lastSuccessfulPoll = pollTime
+	oldestTs := int64(math.MaxInt64)
+	newestTs := int64(0)
 	for _, fieldValue := range fieldValues {
+		if fieldValue.EntityGroupId != dcgm.FE_GPU {
+			continue
+		}
+		gpuIndex := fieldValue.EntityId
+		if _, ok := client.devices[gpuIndex]; !ok {
+			device, err := newDeviceMetrics(client.logger, gpuIndex)
+			if err != nil {
+				continue
+			}
+			client.devices[gpuIndex] = device
+		}
+		device := client.devices[gpuIndex]
 		dcgmName := dcgmIDToName[dcgm.Short(fieldValue.FieldId)]
 		if err := isValidValue(fieldValue); err != nil {
 			msg := fmt.Sprintf("Received invalid value (ts %d gpu %d) %s: %v", fieldValue.Ts, gpuIndex, dcgmName, err)
-			client.issueWarningForFailedQueryUptoThreshold(gpuIndex, dcgmName, msg)
-			if client.retryBlankValues && errors.Is(err, errBlankValue) {
-				retry = true
-			}
+			client.issueWarningForFailedQueryUptoThreshold(fmt.Sprintf("device%d.%s", gpuIndex, dcgmName), msg)
```

Comment:
On H100, we continue getting a lot of these blank value warnings, as late as 5 minutes after startup, until the threshold gets saturated. We are either querying DCGM for these fields too frequently, or there's another reason why blank values are returned. This produces a lot of log spam from the binary — I suggest we suppress blank value errors, at least for the initial ramp-up phase.